src/clj/puppetlabs/services/jruby_pool_manager/impl/jruby_internal.clj

```diff
@@ -86,6 +86,29 @@
    pool :- jruby-schemas/pool-queue-type]
   (.borrowItemWithTimeout pool timeout TimeUnit/MILLISECONDS))
 
+(schema/defn insert-shutdown-poison-pill
+  [pool :- jruby-schemas/pool-queue-type]
+  (.insertPill pool (ShutdownPoisonPill. pool)))
+
+(schema/defn insert-poison-pill
+  [error :- Throwable
+   pool :- jruby-schemas/pool-queue-type]
+  (.insertPill pool (PoisonPill. error)))
+
+(schema/defn fill-pool-with-poison-pills!
+  "Fill an empty pool with poison pills. Should always be run on the
+  modify-instance-agent. insert-fn should be a fn that inserts a poison
+  pill of some kind into the pool"
+  [pool-state :- jruby-schemas/PoolState
+   insert-fn :- IFn]
+  (let [pool (:pool pool-state)
+        pool-size (:size pool-state)]
+    (when (not (= pool-size (.remainingCapacity pool)))
```

Comment:
Consider this sequence:
* File sync requests the lock, blocks
* The pool is flushed - this puts the poison pill into the pool, right?
* As suggested, file sync gets an Exception

That makes plenty of sense, and I can see why you'd like to throw an Exception in that case, but from file sync's perspective it would be better if it could simply keep blocking until the new pool was available and acquire a lock on it, instead of getting an Exception.  That's not a hard requirement - I think it's probably acceptable to say that a flush can prevent code from being sync'ed, but it is a bit unfortunate.  Would that be a new limitation?  Do we know what happens today in this situation?